cheatsheets/Input_Validation_Cheat_Sheet.md

```diff
@@ -166,42 +166,66 @@ The upload feature should be using a whitelist approach to only allow specific f
+- The email address is correct.
+- The application can successfully send emails to it.
+- The user has access to the mailbox.
 
-This does not mean that other users cannot access this mailbox, for example when the user makes use of a service that generates a throw away email address.
+The links that are sent to users to prove ownership should contain a token that is:
 
-- Email verification links should only satisfy the requirement of verify email address ownership and should not provide the user with an authenticated session (e.g. the user must still authenticate as normal to access the application).
-- Email verification codes must expire after the first use or expire after 8 hours if not used.
+- At least 32 characters long.
+- Generated using a [secure source of randomness](Cryptographic_Storage_Cheat_Sheet.md#rule---use-cryptographically-secure-pseudo-random-number-generators-csprng).
+- Single use.
+- Time limited (e.g, expiring after eight hours).
 
-## Address Normalization
+After validating the ownership of the email address, the user should then be required to authenticate on the application through the usual mechanism.
 
-As the local-part of email addresses are, in fact - case sensitive, it is important to store and compare email addresses correctly. To normalise an email address input, you would convert the domain part ONLY to lowercase.
+### Disposable Email Addresses
 
-Unfortunately this does and will make input harder to normalise and correctly match to a users intent. It is reasonable to only accept one unique capitalisation of an otherwise identical address, however in this case it is critical to:
+In some cases, users may not want to give their real email address when registering on the application, and will instead provide a disposable email address. These are publicly available addresses that do not require the user to authenticate, and are typically used to reduce the amount of spam received by users' primary email addresses.
 
-- Store the user-part as provided and verified by user verification.
-- Perform comparisons by `lowercase(provided)==lowercase(persisted)`.
+Blocking disposable email addresses is almost impossible, as there are a large number of websites offering these services, with new domains being created every day. There are a number of publicly available lists and commercial lists of known disposable domains, but these will always be incomplete.
+
+If these lists are used to block the use of disposable email addresses then the user should be presented with a message explaining why they are blocked (although they are likely to simply search for another disposable provider rather than giving their legitimate address).
+
+If it is imperative that disposable email addresses are blocked, then registrations should only be allowed from a specifically whitelisted email providers. However, if this includes public providers such as Google or Yahoo, users can simply register their own disposable address with them.
```

Comment:
- Let's avoid the usage of `imperative` as it might be heavy to non native speakers.
- from ~a~ specifically whitelisted email providers. I believe the `a` should be removed as we are talking about multiple providers.